pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ClientWithSocks5ProxyTest.java

```diff
@@ -0,0 +1,159 @@
+                .socks5ProxyUsername("socks5")
+                .socks5ProxyPassword("pulsar");
+    }
+
+    private void startSocks5Server(boolean enableAuth) {
+        Socks5Config config = new Socks5Config();
+        config.setPort(11080);
+        config.setEnableAuth(enableAuth);
+        server = new Socks5Server(config);
+        Thread thread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    server.start();
+                } catch (Exception e) {
+                    e.printStackTrace();
+                }
+            }
+        });
+        thread.setDaemon(true);
+        thread.start();
+    }
+
+    @AfterMethod(alwaysRun = true)
+    protected void cleanup() throws Exception {
+        internalCleanup();
+        server.shutdown();
+    }
+
+    private void initData() throws PulsarAdminException {
+        admin.tenants().createTenant("public", new TenantInfo() {
+            @Override
+            public Set<String> getAdminRoles() {
+                return Collections.emptySet();
+            }
+
+            @Override
+            public Set<String> getAllowedClusters() {
+                Set<String> clusters = new HashSet<>();
+                clusters.add("test");
+                return clusters;
+            }
+        });
+        admin.namespaces().createNamespace("public/default");
+        admin.topics().createNonPartitionedTopic(topicName);
+    }
+
+    @Test
+    public void testSendAndConsumer() throws PulsarClientException {
+        startSocks5Server(true);
+        // init consumer
+        final String subscriptionName = "socks5-subscription";
+        Consumer<byte[]> consumer = pulsarClient.newConsumer()
+                .topic(topicName)
+                .subscriptionName(subscriptionName)
+                .subscriptionType(SubscriptionType.Shared)
+                .subscribe();
+
+        //init producer
+        Producer<byte[]> producer = pulsarClient.newProducer()
+                .topic(topicName)
+                .create();
+
+        String msg = "abc";
+        producer.send(msg.getBytes());
+        Message<byte[]> message = consumer.receive();
+
+        assertEquals(new String(message.getData()), msg);
+
+        consumer.unsubscribe();
+    }
+
+    @Test
+    public void testDisableAuth() throws PulsarClientException {
+        startSocks5Server(false);
+        ClientBuilder clientBuilder = PulsarClient.builder()
+                .serviceUrl(pulsar.getBrokerServiceUrl())
+                .socks5ProxyAddress(new InetSocketAddress("localhost", 11080));
+        PulsarClient pulsarClient = replacePulsarClient(clientBuilder);
+        Producer<byte[]> producer = pulsarClient.newProducer()
+                .topic(topicName)
+                .create();
+        String msg = "abc";
+        producer.send(msg.getBytes());
+    }
+
+    @Test(timeOut = 5000, expectedExceptions = {ThreadTimeoutException.class})
```

Comment:
Why we will get timeout exception here, because of the pulsar client auto reconnect?

Reply:
This test is for auth enabled with incorrect password.  
If run this test method, pulsar client can't connect to socks5 proxy. The socks5 proxy handler will throw ProxyConnectException, but process result in Netty thread(It's the same behavior with ssl handler). Thus, this will lead pulsar client to reconnect. This is the right behavior, so in order to decrease the time wasted, we'd better stop this with timeout and ThreadTimeoutException.